File: backend/app/schemas/licitacion.py
"""
Pydantic schemas for Licitacion model validation.
"""
from pydantic import BaseModel, Field, field_validator
from typing import Optional
from datetime import datetime, date
from decimal import Decimal


class LicitacionBase(BaseModel):
    """Base licitacion schema with common fields."""

    numero_licitacion: str = Field(..., min_length=1, max_length=100)
    cliente_id: int = Field(..., gt=0)
    nombre_licitacion: str = Field(..., min_length=1, max_length=500)
    descripcion: Optional[str] = None
    objeto_contrato: Optional[str] = None

    # Amounts
    monto_ofertado: Optional[Decimal] = Field(None, ge=0)
    monto_adjudicado: Optional[Decimal] = Field(None, ge=0)
    moneda: str = Field(default="CRC", pattern="^(CRC|USD|EUR)$")

    # Status
    estado_licitacion: str = Field(default="en_preparacion")

    # Important dates
    fecha_publicacion: Optional[date] = None
    fecha_presentacion: Optional[date] = None
    fecha_apertura: Optional[date] = None
    fecha_adjudicacion: Optional[date] = None
    fecha_inicio_contrato: Optional[date] = None
    fecha_fin_contrato: Optional[date] = None

    # Additional information
    tipo_licitacion: Optional[str] = None
    categoria: Optional[str] = None
    prioridad: str = Field(default="media", pattern="^(alta|media|baja)$")

    # Links
    url_portal_compras: Optional[str] = None
    numero_expediente: Optional[str] = Field(None, max_length=100)

    # Guarantees
    monto_garantia_participacion: Optional[Decimal] = Field(None, ge=0)
    fecha_vence_garantia_participacion: Optional[date] = None
    monto_garantia_cumplimiento: Optional[Decimal] = Field(None, ge=0)
    fecha_vence_garantia_cumplimiento: Optional[date] = None

    # Observations
    observaciones: Optional[str] = None
    motivo_rechazo: Optional[str] = None

    @field_validator("estado_licitacion")
    @classmethod
    def validate_estado(cls, v):
        allowed_states = [
            "en_preparacion", "presentada", "en_evaluacion", "adjudicada",
            "en_ejecucion", "finalizada", "desierta", "rechazada"
        ]
        if v not in allowed_states:
            raise ValueError(f"Estado must be one of: {', '.join(allowed_states)}")
        return v

    @field_validator("tipo_licitacion")
    @classmethod
    def validate_tipo(cls, v):
        if v is not None:
            allowed_types = ["publica", "privada", "abreviada", "internacional"]
            if v not in allowed_types:
                raise ValueError(f"Tipo must be one of: {', '.join(allowed_types)}")
        return v

    @field_validator("categoria")
    @classmethod
    def validate_categoria(cls, v):
        if v is not None:
            allowed_categories = ["servicios", "obras", "bienes", "consultoria"]
            if v not in allowed_categories:
                raise ValueError(f"Categoria must be one of: {', '.join(allowed_categories)}")
        return v

    @field_validator("prioridad")
    @classmethod
    def validate_prioridad(cls, v):
        allowed_priorities = ["alta", "media", "baja"]
        if v not in allowed_priorities:
            raise ValueError(f"Prioridad must be one of: {', '.join(allowed_priorities)}")
        return v


class LicitacionCreate(LicitacionBase):
    """Schema for creating a new licitacion."""
    pass


class LicitacionUpdate(BaseModel):
    """Schema for updating a licitacion."""

    numero_licitacion: Optional[str] = Field(None, min_length=1, max_length=100)
    cliente_id: Optional[int] = Field(None, gt=0)
    nombre_licitacion: Optional[str] = Field(None, min_length=1, max_length=500)
    descripcion: Optional[str] = None
    objeto_contrato: Optional[str] = None

    # Amounts
    monto_ofertado: Optional[Decimal] = Field(None, ge=0)
    monto_adjudicado: Optional[Decimal] = Field(None, ge=0)
    moneda: Optional[str] = Field(None, pattern="^(CRC|USD|EUR)$")

    # Status
    estado_licitacion: Optional[str] = None

    # Important dates
    fecha_publicacion: Optional[date] = None
    fecha_presentacion: Optional[date] = None
    fecha_apertura: Optional[date] = None
    fecha_adjudicacion: Optional[date] = None
    fecha_inicio_contrato: Optional[date] = None
    fecha_fin_contrato: Optional[date] = None

    # Additional information
    tipo_licitacion: Optional[str] = None
    categoria: Optional[str] = None
    prioridad: Optional[str] = Field(None, pattern="^(alta|media|baja)$")

    # Links
    url_portal_compras: Optional[str] = None
    numero_expediente: Optional[str] = Field(None, max_length=100)

    # Guarantees
    monto_garantia_participacion: Optional[Decimal] = Field(None, ge=0)
    fecha_vence_garantia_participacion: Optional[date] = None
    monto_garantia_cumplimiento: Optional[Decimal] = Field(None, ge=0)
    fecha_vence_garantia_cumplimiento: Optional[date] = None

    # Observations
    observaciones: Optional[str] = None
    motivo_rechazo: Optional[str] = None
    is_active: Optional[bool] = None

    @field_validator("estado_licitacion")
    @classmethod
    def validate_estado(cls, v):
        if v is not None:
            allowed_states = [
                "en_preparacion", "presentada", "en_evaluacion", "adjudicada",
                "en_ejecucion", "finalizada", "desierta", "rechazada"
            ]
            if v not in allowed_states:
                raise ValueError(f"Estado must be one of: {', '.join(allowed_states)}")
        return v

    @field_validator("tipo_licitacion")
    @classmethod
    def validate_tipo(cls, v):
        if v is not None:
            allowed_types = ["publica", "privada", "abreviada", "internacional"]
            if v not in allowed_types:
                raise ValueError(f"Tipo must be one of: {', '.join(allowed_types)}")
        return v

    @field_validator("categoria")
    @classmethod
    def validate_categoria(cls, v):
        if v is not None:
            allowed_categories = ["servicios", "obras", "bienes", "consultoria"]
            if v not in allowed_categories:
                raise ValueError(f"Categoria must be one of: {', '.join(allowed_categories)}")
        return v


class LicitacionInDB(LicitacionBase):
    """Schema for licitacion in database."""

    licitacion_id: int
    is_active: bool
    is_deleted: bool
    created_at: datetime
    updated_at: Optional[datetime] = None
    created_by: Optional[int] = None
    updated_by: Optional[int] = None

    class Config:
        from_attributes = True


class LicitacionResponse(LicitacionInDB):
    """Schema for licitacion response with computed fields."""

    # Optional: include cliente info
    # cliente: Optional[ClienteResponse] = None


class LicitacionList(BaseModel):
    """Schema for paginated licitacion list."""

    total: int
    items: list[LicitacionResponse]
    page: int
    page_size: int
    total_pages: int


class LicitacionStats(BaseModel):
    """Schema for licitacion statistics."""

    total_licitaciones: int
    por_estado: dict[str, int]
    por_categoria: dict[str, int]
    monto_total_ofertado: Decimal
    monto_total_adjudicado: Decimal
    tasa_exito: float  # porcentaje de adjudicadas

File: frontend/lib/core/models/licitacion.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'licitacion.freezed.dart';
part 'licitacion.g.dart';

@freezed
class Licitacion with _$Licitacion {
  const factory Licitacion({
    @JsonKey(name: 'licitacion_id') required int licitacionId,
    @JsonKey(name: 'numero_licitacion') required String numeroLicitacion,
    @JsonKey(name: 'cliente_id') required int clienteId,
    @JsonKey(name: 'titulo_licitacion') required String tituloLicitacion,
    String? descripcion,
    @JsonKey(name: 'estado_licitacion') required String estadoLicitacion,
    required String categoria,
    @JsonKey(name: 'fecha_publicacion') DateTime? fechaPublicacion,
    @JsonKey(name: 'fecha_presentacion') DateTime? fechaPresentacion,
    @JsonKey(name: 'fecha_adjudicacion') DateTime? fechaAdjudicacion,
    @JsonKey(name: 'monto_estimado') double? montoEstimado,
    @JsonKey(name: 'monto_ofertado') double? montoOfertado,
    @JsonKey(name: 'monto_adjudicado') double? montoAdjudicado,
    @JsonKey(name: 'probabilidad_exito') int? probabilidadExito,
    @JsonKey(name: 'fecha_inicio_contrato') DateTime? fechaInicioContrato,
    @JsonKey(name: 'fecha_fin_contrato') DateTime? fechaFinContrato,
    @JsonKey(name: 'fecha_vencimiento_garantia') DateTime? fechaVencimientoGarantia,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _Licitacion;

  factory Licitacion.fromJson(Map<String, dynamic> json) =>
      _$LicitacionFromJson(json);
}

/// Licitacion status enum
enum LicitacionEstado {
  @JsonValue('en_preparacion')
  enPreparacion('En Preparación', 'preparing'),
  @JsonValue('publicada')
  publicada('Publicada', 'published'),
  @JsonValue('presentada')
  presentada('Presentada', 'submitted'),
  @JsonValue('en_evaluacion')
  enEvaluacion('En Evaluación', 'evaluating'),
  @JsonValue('adjudicada')
  adjudicada('Adjudicada', 'awarded'),
  @JsonValue('en_ejecucion')
  enEjecucion('En Ejecución', 'executing'),
  @JsonValue('finalizada')
  finalizada('Finalizada', 'finished'),
  @JsonValue('rechazada')
  rechazada('Rechazada', 'rejected'),
  @JsonValue('desierta')
  desierta('Desierta', 'deserted');

  const LicitacionEstado(this.displayName, this.key);
  final String displayName;
  final String key;

  static LicitacionEstado fromString(String value) {
    return LicitacionEstado.values.firstWhere(
      (e) => e.name == value || e.key == value,
      orElse: () => LicitacionEstado.enPreparacion,
    );
  }
}

/// Licitacion category enum
enum LicitacionCategoria {
  @JsonValue('servicios')
  servicios('Servicios'),
  @JsonValue('obras')
  obras('Obras'),
  @JsonValue('suministros')
  suministros('Suministros'),
  @JsonValue('consultoria')
  consultoria('Consultoría'),
  @JsonValue('tecnologia')
  tecnologia('Tecnología'),
  @JsonValue('desarrollo')
  desarrollo('Desarrollo');

  const LicitacionCategoria(this.displayName);
  final String displayName;
}

@freezed
class LicitacionList with _$LicitacionList {
  const factory LicitacionList({
    required int total,
    required List<Licitacion> items,
    required int page,
    @JsonKey(name: 'page_size') required int pageSize,
    @JsonKey(name: 'total_pages') required int totalPages,
  }) = _LicitacionList;

  factory LicitacionList.fromJson(Map<String, dynamic> json) =>
      _$LicitacionListFromJson(json);
}
